Document remaining distributions and templatize

rcurtin opened the pull request:
I added documentation for these distributions:

 * `DiscreteGaussianDistribution`
 * `LaplaceDistribution`
 * `GammaDistribution`
 * `RegressionDistribution` (this one took me forever to figure out how to describe)

Then, I added a `MatType` template parameter to each of those distributions and adapted the rest of the code accordingly (this resulted in quite a lot of changes throughout, but they are generally very simple things like `GaussianDistribution` -> `GaussianDistribution<MatType>`).

It may take me a couple iterations to get all of the documentation building right without any errors, but at the very least all of the code and snippets appear to work locally.

My time has been super focused on other things for the last month or two, which is unfortunate, but now that is all clearing up so I am hoping this is the first of many documentation PRs in short succession.

shrit commented on src/mlpack/core/distributions/laplace_distribution.hpp:
```diff
@@ -64,7 +69,7 @@ class LaplaceDistribution
    */
   LaplaceDistribution(const size_t dimensionality, const double scale) :
-      mean(arma::zeros<arma::vec>(dimensionality)), scale(scale) { }
+      mean(arma::zeros<VecType>(dimensionality)), scale(scale) { }
```
```suggestion
      mean(zeros<VecType>(dimensionality)), scale(scale) { }
```
This is supposed to work as I remember

shrit commented on src/mlpack/core/distributions/laplace_distribution.hpp:
```diff
@@ -125,22 +130,12 @@ class LaplaceDistribution
+    VecType result(mean.n_elem);
     result.randu();
-
-    // Convert from uniform distribution to Laplace distribution.
-    // arma::sign() does not exist in Armadillo < 3.920 so we have to do this
-    // elementwise.
-    for (size_t i = 0; i < result.n_elem; ++i)
-    {
-      if (result[i] < 0.5)
-        result[i] = mean[i] + scale * std::log(1 + 2.0 * (result[i] - 0.5));
-      else
-        result[i] = mean[i] - scale * std::log(1 - 2.0 * (result[i] - 0.5));
-    }
-
```
:+1:

shrit commented on src/mlpack/tests/mean_shift_test.cpp:
```diff
@@ -110,32 +110,34 @@ TEMPLATE_TEST_CASE("MeanShiftSimpleCentroidsOnlyTest", "[MeanShiftTest]", float,
     for (size_t i = 0; i < 1000; ++i)
-      dataset.col(i) = arma::conv_to<arma::Col<ElemType>>::from(g1.Random());
+      dataset.col(i) = g1.Random();
     for (size_t i = 1000; i < 2000; ++i)
-      dataset.col(i) = arma::conv_to<arma::Col<ElemType>>::from(g2.Random());
+      dataset.col(i) = g2.Random();
     for (size_t i = 2000; i < 3000; ++i)
-      dataset.col(i) = arma::conv_to<arma::Col<ElemType>>::from(g3.Random());
+      dataset.col(i) = g3.Random();
     for (size_t i = 3000; i < 4000; ++i)
-      dataset.col(i) = arma::conv_to<arma::Col<ElemType>>::from(g4.Random());
+      dataset.col(i) = g4.Random();
```
I have seen these in several places, how are we able to get rid of the conv_to  as we are assigning mat to col?

shrit commented on src/mlpack/core/distributions/laplace_distribution_impl.hpp:
```diff
@@ -49,7 +53,33 @@ inline void LaplaceDistribution::Probability(const arma::mat& x,
  */
-inline void LaplaceDistribution::Estimate(const arma::mat& observations)
+template<typename MatType>
+[[deprecated("Will be removed in mlpack 5.0.0; use Train() instead")]]
+inline void LaplaceDistribution<MatType>::Estimate(const MatType& observations)
+{
+  Train(observations);
+}
+
+/**
+ * Estimate the Laplace distribution directly from the given observations,
+ * taking into account the probability of each observation actually being from
+ * this distribution.
+ */
+template<typename MatType>
+[[deprecated("Will be removed in mlpack 5.0.0; use Train() instead")]]
+inline void LaplaceDistribution<MatType>::Estimate(const MatType& observations,
+                                                   const VecType& probabilities)
+{
+  Train(observations, probabilities);
+}
```
Nice and indeed, does not make sense and better for API compatibility to have Train function everywhere :+1:

shrit reviewed (commented):
Everything looks good to me, but I am a bit confused when it comes to finding the distributions docs. For instance, I had to look at the core.md to understand where to look in the doc page.

The main problem is that the core page is not referenced in the main page, it is true we can find it in the mlpack basic, but I doubt why anyone would try to look in there.

We need to provide more initiative way, maybe adding a tree based sub elements on the left side showing what is inside the core page ? this could help to understand what are the core functionality in the library without going there?

Also it would be nice to reduce core library functions to core functions, library seems to be a bit redundant, or maybe change core to something that describes more what the code is doing instead of its location ?

I would also move aliases from that page to the developer page, since I doubt the user would ever need to create an alias of any matrix.

I would also move kernel and distribution to its own page, because they are too long in here and for the user it makes more sense following what we have did previously for different methods.

Also, I have noticed that some of the distributions do not have definition like the others;
![Screenshot 2024-09-01 at 06-32-49 ](https://github.com/user-attachments/assets/182ec7b7-6e47-4b77-8f2d-710b8a5a4417)


Apart from that, the internal docs of the distributions look perfect, I just need to review the test for the distribution after we address these comments

rcurtin commented:
> Everything looks good to me, but I am a bit confused when it comes to finding the distributions docs. For instance, I had to look at the core.md to understand where to look in the doc page.
> 
> The main problem is that the core page is not referenced in the main page, it is true we can find it in the mlpack basic, but I doubt why anyone would try to look in there.
> 
> We need to provide more initiative way, maybe adding a tree based sub elements on the left side showing what is inside the core page ? this could help to understand what are the core functionality in the library without going there?
>
> Also it would be nice to reduce core library functions to core functions, library seems to be a bit redundant, or maybe change core to something that describes more what the code is doing instead of its location ?
>
> I would also move aliases from that page to the developer page, since I doubt the user would ever need to create an alias of any matrix.
> 
> I would also move kernel and distribution to its own page, because they are too long in here and for the user it makes more sense following what we have did previously for different methods.

Yes, I think that you are right.  For now I committed 7ef9665 which is a little bit of a workaround---it adds the subsections from `core.md` to the sidebar.  I also changed `Core library functions` to `Core functionality` like you suggested.

I think that `core.md` should be split up.  It seems like a better way would be to split it into a structure like this:

 * `core/math.md`
 * `core/distances.md`
 * `core/distributions.md`
 * `core/kernels.md`
 * `core/trees.md` (although there are so many trees, each one might need its own page...)

What do you think?  If that's a reasonable structure I'll open a separate PR for it shortly (a separate PR so that the diff is way easier to read).

> Also, I have noticed that some of the distributions do not have definition like the others

Thank you!  I overlooked that.  Fixed in bc907953.

> Apart from that, the internal docs of the distributions look perfect, I just need to review the test for the distribution after we address these comments

:+1:

shrit commented:
Perfect, make sense, to avoid the merge conflict, I will approve this so we can merge it and then merge #3792

shrit commented on src/mlpack/tests/distribution_test.cpp:
```diff
@@ -314,85 +447,99 @@ TEST_CASE("GaussianDistributionProbabilityTest", "[DistributionTest]")
+  typedef typename arma::Col<ElemType> VecType;
+  typedef typename arma::Mat<ElemType> MatType;
+
+  const ElemType tol = (std::is_same<ElemType, float>::value) ? 1e-4 : 1e-7;
```
:+1:
